theano/sandbox/jaxify.py

```diff
@@ -462,63 +468,74 @@ def jax_args_to_inner_scan(op, carry, x):
             inner_scan_inputs = [
                 inner_in_seqs,
-                inner_in_mit_mot,
-                inner_in_mit_sot,
+                # inner_in_mit_mot,
+                inner_in_mit_sot_flatten,
                 inner_in_sit_sot,
+                inner_in_shared,
                 inner_in_non_seqs,
             ]
 
-            raise NotImplementedError()
             return inner_scan_inputs
 
         def inner_scan_outs_to_jax_outs(
             op,
             old_carry,
             inner_scan_outs,
         ):
-            # `inner_scan_outs` is a list with the following
-            # composite form:
-            # outer_out_mit_mot
-            # + outer_out_mit_sot
-            # + outer_out_sit_sot
-            # + outer_out_nit_sot
-            # + outer_out_shared
-            # + cond
             (
-                outer_out_mit_mot,
                 outer_out_mit_sot,
-                outer_out_sit_sot,
-                outer_out_nit_sot,
-                outer_out_shared,
-                cond,
-            ) = inner_scan_outs
-            outer_out_non_seqs = old_carry[:-n_non_seqs]
+                outer_in_sit_sot,
+                inner_in_shared,
+                outer_in_non_seqs,
+            ) = old_carry
+            if not outer_out_mit_sot:
+                outer_out_mit_sot_next = []
+            else:
+                outer_out_mit_sot_next = jax.tree_multimap(
+                    lambda mit_sot, new_val: jnp.concatenate(
+                        [mit_sot[1:], new_val[None, ...]], axis=0
+                    ),
+                    outer_out_mit_sot,
+                    inner_scan_outs,
+                )
 
             # This should contain all inner-output taps, non_seqs, and shared
             # terms
-            carry = [
-                outer_out_mit_mot,
-                outer_out_mit_sot,
-                outer_out_sit_sot,
-                outer_out_shared,
-                outer_out_non_seqs,
-            ]
-            # This should contain all inner-outputs that produce
-            # outer-outputs
-            y = []
+            if not outer_in_sit_sot:
+                outer_in_sit_sot_next = []
+            else:
+                outer_in_sit_sot_next = inner_scan_outs
+            new_carry = (
+                outer_out_mit_sot_next,
+                outer_in_sit_sot_next,
+                inner_in_shared,
+                outer_in_non_seqs,
+            )
 
-            raise NotImplementedError()
-            return (carry, y)
+            return new_carry
 
         def jax_inner_func(carry, x):
             inner_args = jax_args_to_inner_scan(op, carry, x)
-            inner_scan_outs = jax_tt_inner_func(*inner_args)
-            new_carry, y = inner_scan_outs_to_jax_outs(op, inner_scan_outs)
-            return new_carry, y
-
-        return jax.lax.scan(jax_inner_func, init_carry, seqs, length=n_steps)
+            inner_args = jax.tree_leaves(inner_args)
```

Comment:
Is `jax.tree_leaves(inner_args)` effectively equivalent to `sum(inner_args, [])`?

Reply:
Yes.